axelor-sale/src/main/java/com/axelor/apps/sale/service/saleorder/SaleOrderPackagingServiceImpl.java

```diff
@@ -69,10 +70,52 @@ public String estimatePackaging(SaleOrder saleOrder) throws AxelorException {
                 Collectors.toMap(
                     SaleOrderLine::getProduct, SaleOrderLine::getQty, BigDecimal::add));
 
-    List<Product> products = getSortedProducts(productQtyMap);
     List<String> messages = new ArrayList<>();
-    List<Product> boxes = productRepository.all().filter("self.isPackaging = true").fetch();
-    boxes.sort(Comparator.comparing(this::getBoxInnerVolume));
+    List<Product> packagingOptions =
+        productRepository.all().filter("self.isPackaging = true").fetch();
+
+    Map<Integer, Map<Product, String>> levelDescriptions = new HashMap<>();
+    Map<Integer, Map<Product, BigDecimal[]>> levelWeights = new HashMap<>();
+
+    packByLevel(
+        ProductRepository.PACKAGING_LEVEL_BOX,
+        packagingOptions,
+        productQtyMap,
+        messages,
+        levelDescriptions,
+        levelWeights);
+
+    if (CollectionUtils.isEmpty(messages)) {
+      return "";
+    }
+    return StringHtmlListBuilder.formatMessage(
+        I18n.get("Packaging") + " " + I18n.get(saleOrder.getFullName()), messages);
+  }
+
+  protected void packByLevel(
+      int level,
+      List<Product> packagingOptions,
+      Map<Product, BigDecimal> productQtyMap,
+      List<String> messages,
+      Map<Integer, Map<Product, String>> levelDescriptions,
+      Map<Integer, Map<Product, BigDecimal[]>> levelWeights)
+      throws AxelorException {
+
```

Comment:
Please add a security to prevent infinite loop. For example if the level exceeds a number of iterations. Similar to SupplychainExceptionMessage.MRP_TOO_MANY_ITERATIONS.